### ISSUE TITLE
ansible.builtin.user: when parent of home is non existent a created home may have wrong permissions

### ISSUE THREAD

**Issue by flybyray** [ansible/ansible#77260]
Added minimal reproducible code and updated this issue in comment: https://github.com/ansible/ansible/issues/77260#issuecomment-1066724650

This original summary was based on wrong assumptions and hence contains wrong conclusions:

### Summary

User home directories created by `ansible.builtin.user`  module will be created with `/etc/login.defs`  `UMASK` setting. Not checking if `HOME_MODE`  is set.
Hence the permissions are not what you would expect if HOME_MODE is set.

refs:

* relevant code which produces this issue: https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/user.py#L1259-L1269
* description and defaults from shadow: https://github.com/shadow-maint/shadow/blob/master/etc/login.defs#L195-L207

Please consider this might even be a security issue!

### Issue Type

Bug Report

### Component Name

user

### Ansible Version

```console
$ ansible --version


ansible [core 2.12.3]
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/robert.rettig/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/robert.rettig/stages/test/venv/lib64/python3.10/site-packages/ansible
  ansible collection location = /home/robert.rettig/.ansible/collections:/usr/share/ansible/collections
  executable location = /home/robert.rettig/stages/test/venv/bin/ansible
  python version = 3.10.2 (main, Jan 17 2022, 00:00:00) [GCC 11.2.1 20211203 (Red Hat 11.2.1-7)]
  jinja version = 3.0.3
  libyaml = True
```
```


### Configuration

```console
$ ansible-config dump --only-changed # is empty


```
```


### OS / Environment

* EL8
    * AlmaLinux 8.5 (Arctic Sphynx)
* Fedora Linux 35 (Thirty Five)


### Steps to Reproduce

<!--- Paste example playbooks or commands between quotes below -->
```yaml (paste below)
- ansible.builtin.user:
    create_home: True
    name: "nonexistentuser"
```

### Expected Results

The permissions for a created home directory should be the same as if the home directory had been created with 
`useradd` or `newusers`.

Home dir needs to be created with `HOME_MODE`.

```shell
$ grep -E '^UMASK|^HOME_MODE' /etc/login.defs 
UMASK           022
HOME_MODE       0700

$ sudo useradd testuser
$ ls -lad ~testuser
drwx------. 2 testuser testuser 62 Mar 11 17:00 /home/testuser
```

### Actual Results

```console
$ grep -E '^UMASK|^HOME_MODE' /etc/login.defs 
UMASK           022
HOME_MODE       0700

$ ls -lad ~nonexistentuser
drwxr-xr-x. 17 nonexistentuser nonexistentuser 4096 Mar 10 17:41 /home/nonexistentuser
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

**Comment by ansibot** [comment]
Files identified in the description:
* [`lib/ansible/modules/user.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/user.py)

If these files are incorrect, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/devel/ISSUE_HELP.md)

**Comment by sysadmin75** [comment]
I wasn't able to reproduce this issue. I tested on an AlmaLinux 8.5 server.

```
[root@almalinux ~]# grep -E '^UMASK|^HOME_MODE' /etc/login.defs 
UMASK		022
HOME_MODE	0700
[root@almalinux ~]# useradd -m -d /home/deleteme deleteme
[root@almalinux ~]# ll /home/deleteme/ -d
drwx------ 2 deleteme deleteme 62 Mar 13 14:17 /home/deleteme/
[root@almalinux ~]#
```
My task is defined as so:
```
$ cat createuser.yml 
---
- hosts: alma
  become: False
  remote_user: root
  tasks:
    - name: Create user account
      ansible.builtin.user:
        create_home: True
        name: "nonexistentuser"
```
Running the task:
```
$ ansible-playbook -i hosts createuser.yml 

PLAY [alma] **********************************************************************************************************************************************************************************

TASK [Gathering Facts] ***********************************************************************************************************************************************************************
[WARNING]: Platform linux on host alma is using the discovered Python interpreter at /usr/bin/python3.6, but future installation of another Python interpreter could change the meaning of
that path. See https://docs.ansible.com/ansible-core/2.12/reference_appendices/interpreter_discovery.html for more information.
ok: [alma]

TASK [Create user account] *******************************************************************************************************************************************************************
changed: [alma]

PLAY RECAP ***********************************************************************************************************************************************************************************
alma                       : ok=2    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```
And the permissions are the same as when I created the `deleteme` user above.
```
[root@almalinux ~]# ll /home/nonexistentuser/ -d
drwx------ 2 nonexistentuser nonexistentuser 62 Mar 13 14:33 /home/nonexistentuser/
[root@almalinux ~]#
```

**Comment by flybyray** [comment]
Hi @sysadmin75 thanks for looking into this.

I guess you did not reproduced it as i described it. The problem is as referenced by this change: https://github.com/ansible/ansible/commit/a11bb8b4d39b1ff77762d58ffea42eca63f6f5f5

You did not supplied the ansible version for your try-out. You really need to ensure to have this output on your controller

```
$ ansible --version

ansible [core 2.12.3]
```

**Comment by flybyray** [comment]
@sysadmin75 sorry i need to sort this out. i tried to make it reprocuable with podman/docker. but really i cannot reproduce it there. 
I tried to extract the problem out of a bigger ansible playbook. Which works on almalinux/8 libvirt image.
I thought I found the minimal,  but as i now see with  docker/podman  i cannot reproduce it too. Hence I will close this for now.

In the above issue some formatting issue happened:

![grafik](https://user-images.githubusercontent.com/804661/158135792-4e4942c8-972c-48a1-b5dd-aca8964cfb3f.png)

Something might be very relevant from the playbook was cut off.  I used the new parameter umask ( ansible-core 2.12 ) too with a value "077". 
https://docs.ansible.com/ansible/latest/collections/ansible/builtin/user_module.html#parameter-umask

But even with using that playbook it is not reproducible in docker/podman. Hence I will close this for now. Might be I will reopen it if I find the reason ( maybe it is something with the libvirtd almalinux/8 image )

**Comment by flybyray** [comment]
it is really crazy what is going on there i think i need to debug it somehow on the module level on the host itself.
i just noticed that the user module writes a warning to stderr

```txt
"stderr": "useradd: warning: the home directory already exists.\\nNot copying any file from skel directory into it.\\n"
```

And i added directly in front a stat moduke check on that directory to prove it was not created before the module call.

Hence something is going on in the user module which tries to create the homedirectory twice?

**Comment by flybyray** [comment]
@sysadmin75 and @mkrizek can you please make a suggestion?

Should I create a new issue? Or edit the issue head?
For now i just reopen. 

Here: https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/user.py#L701
it should create the `parent` and not the `self.home` itself.
Because of that I receive:

```
"stderr": "useradd: warning: the home directory already exists.\\nNot copying any file from skel directory into it.\\n"
```

an the home is created with wrong permissions. In my playbook I use

```yaml
- name: user | ensure mysql user is set
  ansible.builtin.user:
    comment: "user for instance {{ _instance_name }}"
    create_home: True
    generate_ssh_key: yes
    group: "my-{{ _instance_name }}"
    groups:
      - systemd-journal
    home: "/path/to/nonexistentparent/{{ _instance_name }}"
    name: "my-{{ _instance_name }}"
    shell: /bin/bash
    system: true
    uid: "33000"
    umask: "077"
```

The wrong home dir permissions problem shows up for the first instance on each host.
I think this is very obvious a bug.

**Comment by jborean93** [comment]
Hi!

Thanks very much for your submission to Ansible.  It means a lot to us that you've taken time to contribute.

Unfortunately, we're not sure if we want this feature in the program, and I don't want this to seem confrontational.
Our reasons for this are:

   * This is not something we are planning on implementing
   * If you have special requirements it should be done before calling this module, [file](https://docs.ansible.com/ansible/latest/collections/ansible/builtin/file_module.html) is a good candidate

However, we're absolutely always up for discussion.
Because this project is very active, we're unlikely to see comments made on closed tickets and we lock them after some time.
If you or anyone else has any further questions, please let us know by using any of the communication methods listed in the page below:

   * https://docs.ansible.com/ansible/latest/community/communication.html

In the future, sometimes starting a discussion on the development list prior to implementing a feature can make getting things included a little easier, but it's not always necessary.

Thank you once again for this and your interest in Ansible!
